### ISSUE TITLE
Support recompiling and reloading Java classes better

### ISSUE THREAD

**Issue by RadicalZephyr** [boot-clj/boot#522]
This is just a place to capture some discussion around the idea of adding support for reloading Java classes into a running repl, something like how [virgil](https://github.com/ztellman/virgil) does.

As an initial first pass, [this new javac task](https://github.com/RadicalZephyr/boot-junit/blob/temp-javac-task/src/radicalzephyr/boot_junit.clj#L119-L155) can be used to get pure Java code to work with `watch` recompiling.

Questions to address:

How much of what Virgil does is necessary to do within boot?  Clearly the file watching isn't needed, but it's not clear what needs to be done regarding namespace reloading.

The POC reloading `javac` task linked above only produces the class file in memory, meaning that the user needs to choose between producing file artifacts and being able to reload the classes.  This is not ideal, but I'm not clear on whether both artifacts could be produced without issues.

Related to the fact that this works in memory is that this strategy won't work across Pod boundaries (I think). Again, more investigation needed.

**Comment by alandipert** [comment]
Did you find namespace reloading necessary in the case where you had Clojure code that relied on dynamically-recompiled Java classes?

I couldn't tell just from reading the virgil code whether or not this was necessary, or served some other purpose.

All in all though, this is a great idea, and would be great to support somehow.

**Comment by stephenbrady** [comment]
I wrote a version of this as well based on the virgil code (which has many correctness issues by the way).

Namespace reloading isn't necessary per se because part of that code backdoors into `clojure.lang.DynamicClassLoader` and mutates that classloader's class cache. If you still have instances of those overwritten classes hanging around though, obviously you'll have that brokenness just like with modifying a `deftype`.

On the file-watching point, it is actually needed if you want to have a Clojure repl and boot-controlled watch task. It's why I created this issue: https://github.com/boot-clj/boot/issues/492

On producing artifacts, you can always dump out the OutputStream with the class in memory to the `target` directory just like `javac` would and have both in sync.

Would also love to see this brought into boot itself, but I think https://github.com/boot-clj/boot/issues/492 has to be addressed to be practical. That issue describes one clean approach, but in the interim, I've been monkey-patching around `boot.core/src-watcher`.

**Comment by micha** [comment]
@stephenbrady Did you find any cases where you needed to use the simpler built-in task, like compiling for production, uberjars, etc? Or can the reloadable `javac*` implementation be used as a drop-in replacement for the built-in `javac` task? Is it backward compatible?

**Comment by micha** [comment]
Also does it work with Java 1.7+?

**Comment by RadicalZephyr** [comment]
FYI: I've put my version of the `javac*` task into it's own repo now: https://github.com/RadicalZephyr/boot-javac-star

I'll be experimenting with the concept there, and suggestions for changes (or PRs) are most welcome.

**Comment by stephenbrady** [comment]
@micha Yes, we do use the current `javac` boot task for producing an uberjar even though we don't use that task while in repl mode. I do think it's needed as a separate pipeline task because I would assume the common case is *not* doing dynamic class loading (of Java classes) in production. No opinion here as to whether the boot `javac` task should wrap both the static and dynamic compilation case, but internally, the implementation of the two cases will have to be different. The dynamic case in a Clojure repl requires reaching into `clojure.lang.DynamicClassLoader`; the static case does not. So no, I don't think it'd be a good idea to treat the dynamic implementation as a replacement for the current static `javac` task.

Not sure what you're hinting at with Java 1.7+. This setup is working for us in Java 1.8. Haven't tried earlier Java versions, but I also don't know why it wouldn't work.

**Comment by alexander-yakushev** [comment]
Boot support has just landed into [virgil](https://github.com/ztellman/virgil). At the same time, I doubt it makes sense to reimplement such functionality in Boot natively since it is actually quite a lot of work. It seems trivial, but then you need to handle annotations, properly order the loading of classes dependency-wise, etc. I think this issue can be closed now. @RadicalZephyr?

**Comment by RadicalZephyr** [comment]
Sounds good to me. A boot task for reloading  java classes using virgil
definitely seems like a better approach than including it in core boot.

For my own needs (testing java classes) I’ve found that pod isolation was
all that’s needed.
-- 
Geoff (from mobile) Nothing is ever easy.
